Calculator/Calculator.cpp

```diff
@@ -32,100 +68,278 @@ class Token_stream {
 
-Token_stream::Token_stream()   
-    :full(false), buffer(0)    
+Token_stream::Token_stream()
+    :full(false), buffer(0)
 {
 }
 
-//------------------------------------------------------------------------------
+//-----------------------------------------------------------------------------
+
+const char number = 'n';
+const char quit = 'q';
+const char print = ';';
+const string prompt = ">";
+const string result = "=";
+const char name = 'a';
+const char let = 'L';
+const string declkey = "let";
+const char root = 's';
+const string sqrtkey = "sqrt";
+const char power = 'p';
+const string powkey = "pow";
 
-void Token_stream::putback(Token t)
-{
-    if (full) error("putback() into a full buffer");
-    buffer = t; 
-    full = true; //Buffer cannot be overwritten
-}
 
 //------------------------------------------------------------------------------
-// Evaluates the buffer and assembles and returns Token(ch) or Token(ch,val) depending 
+// Evaluates the buffer and assembles and returns Token(ch), Token(ch,val), Token  or Token depending 
 // on character input
 
 Token Token_stream::get()
 {
-    if (full) { 
+    if (full) {
         full = false; // Buffer can be overwritten
         return buffer; // Returns token currently in buffer
     }
     char ch;
-    cin >> ch; 
+    cin >> ch;
     switch (ch) {
-        case '=': 
-        case 'x': 
-        case '!': 
-        case '{': case '}': case '(': case ')': case '+': case '-': case '*': case '/':
+        case quit:
+        case print:
+        case '!':
+        case '{':
+        case '}':
+        case '(':
+        case ')':
+        case '+':
+        case '-':
+        case '*':
+        case '/':
+        case '%':
+        case '=':
+        case ',':
         {
-            return Token{ch}; 
+            return Token{ ch };
         }
-        case '.':
-        case '0': case '1': case '2': case '3': case '4':
+        case '.': case '0': case '1': case '2': case '3': case '4':
         case '5': case '6': case '7': case '8': case '9':
-        { 
-            cin.putback(ch); 
+        {
+            cin.putback(ch);
             double val;
-            cin >> val; 
-            return Token{'n',val}; 
+            cin >> val;
+            return Token{ number,val };
         }
-        default:
+        default: 
         {
-            error("Bad token"); 
-        } 
+            if (isalpha(ch)) {
+                string s;
+                s += ch;
+                while (cin.get(ch) && (isalpha(ch) || isdigit(ch))) {
+                    s += ch;
+                }
+                cin.putback(ch);
+                if (s == declkey) {
+                    return Token{ let };
+                }
+                if (s == sqrtkey) {
+                    return Token{ root };
+                }
+                if (s == powkey) {
+                    return Token{ power };
+                }
+                return Token{ name, s };
+            }
```

Comment:
what's going on here?

Reply:
it allows the reading of alpha character strings put in by the user so that they can return tokens that can be used to call different functions. for example if the user types "sqrt(9);" it will compare "sqrt" to my square root "key" and return the "root" token, which is caught in primary() and calls my squareroot() function on whatever expression is in the parenthesis.